tokio/src/sync/task/atomic_waker.rs

```diff
@@ -178,8 +182,16 @@ impl AtomicWaker {
         {
             WAITING => {
                 unsafe {
+                    // If `into_waker` panics (because it's code outside of
+                    // AtomicWaker) we need to prime a guard that is called on
+                    // unwind to restore the waker to a WAITING state. Otherwise
+                    // any future calls to register will incorrectly be stuck
+                    // believing it's being updated by someone else.
+                    let guard = PanicGuard(&self.state);
+                    let new_waker = waker.into_waker();
+                    std::mem::forget(guard);
                     // Locked acquired, update the waker cell
-                    self.waker.with_mut(|t| *t = Some(waker.into_waker()));
+                    self.waker.with_mut(|t| *t = Some(new_waker));
```

Comment:
I just realized that this can panic too, since the old waker may have a destructor.